Source/Events.Handling/Builder/ConventionEventHandlerBuilder.cs

```diff
@@ -0,0 +1,319 @@
+                    shouldAddHandler = false;
+                }
+
+                if (!ParametersAreOkay(method, eventHandlerId, out var parameterWarnings))
+                {
+                    warningMessages.AddRange(parameterWarnings.Warnings);
+                    shouldAddHandler = false;
+                }
+
+                if (shouldAddHandler) AddTypedHandleSignatureToBuilder(eventParameterType, method, eventHandlerMethodsBuilder);
+            }
+
+            if (warningMessages.Count == 0) return true;
+
+            warnings = new EventHandlerBuildWarnings(eventHandlerId, warningMessages);
+            return false;
+        }
+
+        bool ParametersAreOkay(MethodInfo method, EventHandlerId eventHandlerId, out EventHandlerBuildWarnings buildWarnings)
+        {
+            buildWarnings = default;
+            var warnings = new List<string>();
+            if (!SecondMethodParameterIsEventContext(method))
+            {
+                warnings.Add($"Event handler method {method} on event handler {_eventHandlerType} needs to have two parameters where the second parameter is {typeof(EventContext)}");
+            }
+
+            if (!MethodHasNoExtraParameters(method))
+            {
+                warnings.Add($"Event handler method {method} on event handler {_eventHandlerType} needs to only have two parameters where the first is the event to handle and the second is {typeof(EventContext)}");
+            }
+
+            if (MethodReturnsAsyncVoid(method) || (!MethodReturnsVoid(method) && !MethodReturnsTask(method)))
+            {
+                warnings.Add($"Event handler method {method} on event handler {_eventHandlerType} needs to return either {typeof(void)} or {typeof(Task)}");
+            }
+
+            if (warnings.Count == 0) return true;
+            buildWarnings = new EventHandlerBuildWarnings(eventHandlerId, warnings);
+            return false;
+        }
+
+        void AddUntypedHandleSignatureToBuilder(MethodInfo method, EventType eventType, EventHandlerMethodsBuilder builder)
+        {
+            var eventHandlerSignatureType = method.ReturnType == typeof(Task) ? typeof(EventHandlerSignature) : typeof(VoidEventHandlerSignature);
+            var eventHandlerSignature = Delegate.CreateDelegate(eventHandlerSignatureType, _eventHandlerInstance, method);
+            var builderHandleMethod = typeof(EventHandlerMethodsBuilder)
+                                        .GetMethod(nameof(EventHandlerMethodsBuilder.Handle), new[] { typeof(EventType), eventHandlerSignatureType });
+            if (eventHandlerSignature is EventHandlerSignature signature) builder.Handle(eventType, signature);
+            else if (eventHandlerSignature is VoidEventHandlerSignature voidSignature) builder.Handle(eventType, voidSignature);
```

Comment:
Seems like there is a chance `Delegate.CreateDelegate` can return null and nothing will be added silently here? Also the if and the first ternary seem to check for the same thing, perhaps we should do the check with a boolean and share that for both of them?

Reply:
This should be resolved by latest commits